mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -200,15 +199,24 @@ impl StreamWakeableState for RingBuffer {
         let block_size = *SERIALIZED_BLOCK_SIZE;
         let total_size = block_size + data_size;
 
-        // Check that we have enough space to insert data.
         // If we have set can_read_from_wrap_around_when_write_full
         // then we must also be full.
-        let free_space = (read_index + self.max_file_size - write_index) % self.max_file_size;
         if self.metadata.can_read_from_wrap_around_when_write_full {
             return Err(PersistError::RingBuffer(RingBufferError::Full));
         }
-        if read_index != write_index && free_space < total_size {
-            return Err(PersistError::RingBuffer(RingBufferError::Full));
+
+        // Check that we have enough space to insert data.
+        let mut free_space = read_index + self.max_file_size - write_index;
+        if write_index < read_index {
+            free_space %= self.max_file_size;
+        }
+        if free_space < total_size {
+            return Err(PersistError::RingBuffer(
+                RingBufferError::InsufficientSpace {
+                    free: free_space,
+                    required: total_size,
+                },
+            ));
```

Comment:
Just to make clear the issue for anyone else:
I thought that std File set_len was a max bound, but it is not.. from when I was using the MMAP crate it did work this way and an err would happen. What this change does is it makes sure that if we have both pointers at the same spot and no wrap around we calculate the remaining size correctly.

E.G. buffer is 200 and both pointers are 100 -> 100 + 200 - 100 => 200 now where before the result would be 0.